### PR TITLE
fix 2017 port `BinaryEditor`

### LINKED ISSUE #2017: Port `System.ComponentModel.Design.BinaryEditor`
**Problem description:**

`BinaryEditor` from System.Design is missing.

Refer to https://github.com/dotnet/winforms/issues/1115#issuecomment-501115599

### PR DESCRIPTION
**Pull request by RussKie**
Port BinaryEditor from NET Fx

Closes #2017

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/2123)

### CODE REVIEW

File: .gitattributes [hunk 1]
```diff
@@ -1 +1,4 @@
 * text=auto
+
+# Collapse XLF files in PRs by default
+*.xlf linguist-generated=true
```
Comment (zsd4yr) [inline .gitattributes]: I like this change but I think it should be made in its own PR or at least its own commit

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/BinaryEditor.cs [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+        }
+
+        /// <summary>
+        ///  Converts the given object to an array of bytes to be manipulated
+        ///  by the editor.  The default implementation of this supports
+        ///  byte[] and stream objects.
+        /// </summary>
+        internal byte[] ConvertToBytes(object value)
+        {
+            if (value is Stream)
```
Comment (zsd4yr) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/BinaryEditor.cs]: nit: you could clean this up a little using an type-style switch-statement

```c#
switch (value)
{
    case Stream stream:
        stream.Position = 0
        int byteCount = (int)(stream.Length - stream.Position);
        byte[] bytes = new byte[byteCount];
        stream.Read(bytes, 0, byteCount);
        return bytes;

    case byte[] bytes:
        return bytes;

    case string str:
        int size = str.Length * 2;
        byte[] buffer = new byte[size];
        Encoding.Unicode.GetBytes(((string)value).ToCharArray(), 0, size / 2, buffer, 0);
        return buffer;
}

Debug.Fail("No conversion from " + value == null ? "null" : value.GetType().FullName + " to byte[]");
return null;
```

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/BinaryEditor.cs [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+            }
+            _binaryUI.Value = null;
+
+            return value;
+        }
+
+        /// <summary>
+        ///  Retrieves the editing style of the Edit method.  If the method is not supported, this will return None.
+        /// </summary>
+        [SuppressMessage("Microsoft.Security", "CA2123:OverrideLinkDemandsShouldBeIdenticalToBase")] // everything in this assembly is full trust.
```
Comment (zsd4yr) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/BinaryEditor.cs]: I think you can remove these?

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/BinaryEditor.cs [hunk 4]
```diff
@@ -0,0 +1,441 @@
...
+        public override UITypeEditorEditStyle GetEditStyle(ITypeDescriptorContext context)
+        {
```
Comment (zsd4yr) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/BinaryEditor.cs]: nit: cleanup in favor of new language features

```c#
public override UITypeEditorEditStyle GetEditStyle(ITypeDescriptorContext context)
    => UITypeEditorEditStyle.Modal;
```

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/BinaryEditor.cs [hunk 5]
```diff
@@ -0,0 +1,441 @@
...
+            {
+                helpService.ShowHelpFromKeyword(HELP_KEYWORD);
+            }
+            else
+            {
+                Debug.Fail("Unable to get IHelpService.");
+            }
+        }
+
+        private class BinaryUI : Form
```
Comment (zsd4yr) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/BinaryEditor.cs]: Please separate this to its own script BinaryEditor.BinaryUI and make BinaryEditor partial to hold it

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/BinaryEditor.cs [hunk 6]
```diff
@@ -0,0 +1,441 @@
...
+        {
+            private BinaryEditor editor;
```
Comment (zsd4yr) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/BinaryEditor.cs]: these privates do not conform to naming convention

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/BinaryEditor.cs [hunk 7]
```diff
@@ -0,0 +1,441 @@
...
+            }
+
+            private void ButtonSave_click(object source, EventArgs e)
+            {
+                try
+                {
+                    SaveFileDialog sfd = new SaveFileDialog();
+                    sfd.FileName = SR.BinaryEditorFileName;
+                    sfd.Title = SR.BinaryEditorSaveFile;
+                    sfd.Filter = SR.BinaryEditorAllFiles + " (*.*)|*.*";
```
Comment (zsd4yr) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/BinaryEditor.cs]: please separate magic string to const

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/BinaryEditor.cs [hunk 8]
```diff
@@ -0,0 +1,441 @@
...
+                catch (IOException x)
+                {
+                    RTLAwareMessageBox.Show(null, string.Format(SR.BinaryEditorFileError, x.Message),
+                                    SR.BinaryEditorTitle, MessageBoxButtons.OK,
+                                    MessageBoxIcon.Error, MessageBoxDefaultButton.Button1, 0);
+                }
+            }
+
+            private void Form_HelpRequested(object sender, HelpEventArgs e)
+            {
```
Comment (zsd4yr) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/BinaryEditor.cs]: cleanup with lambda please

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ByteViewer.cs [hunk 9]
```diff
@@ -0,0 +1,753 @@
...
+namespace System.ComponentModel.Design
+{
+    /// <summary>
+    ///  Displays byte arrays in HEXDUMP, ANSI and Unicode formats.
+    /// </summary>
+    [ToolboxItem(false)]
+    [DesignTimeVisible(false)]
+    public class ByteViewer : TableLayoutPanel
+    {
+        private const int DEFAULT_COLUMN_COUNT = 16;
```
Comment (zsd4yr) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ByteViewer.cs]: consts to not conform to naming convention

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ByteViewer.cs [hunk 10]
```diff
@@ -0,0 +1,753 @@
...
+        private const int DUMP_WIDTH = CHAR_WIDTH * COLUMN_COUNT;
+        private const int HEX_DUMP_GAP = 5;
+        private const int ADDRESS_START_X = BORDER_GAP + INSET_GAP;
+        private const int CLIENT_START_Y = BORDER_GAP + INSET_GAP;
+        private const int LINE_START_Y = CLIENT_START_Y + CELL_HEIGHT / 8;
+        private const int HEX_START_X = ADDRESS_START_X + ADDRESS_WIDTH;
+        private const int DUMP_START_X = HEX_START_X + HEX_WIDTH + HEX_DUMP_GAP;
+        private const int SCROLLBAR_START_X = DUMP_START_X + DUMP_WIDTH + HEX_DUMP_GAP;
+
+        private static readonly Font ADDRESS_FONT = new Font("Microsoft Sans Serif", 8.0f);
```
Comment (zsd4yr) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ByteViewer.cs]: would nameof(FontFamily.SanSerif) work?

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ByteViewer.cs [hunk 11]
```diff
@@ -0,0 +1,753 @@
...
+        private static readonly Font HEXDUMP_FONT = new Font("Courier New", 8.0f);
+
+        private int SCROLLBAR_HEIGHT;
```
Comment (zsd4yr) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ByteViewer.cs]: name of privates does not conform to naming convention

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ByteViewer.cs [hunk 12]
```diff
@@ -0,0 +1,753 @@
...
+            //Assign an index (label) value for 3rd and 4th byte            
+            c5 = GetEncodingIndex(c2);
+
+            //Bellow table is to identify Encoding type based on
+            //first four bytes, those we have converted in index 
+            //values for this look up table
+            //values on column are first two bytes and
+            //values on rows are 3rd and 4th byte 
+
+            int[,] encodings = {
```
Comment (zsd4yr) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ByteViewer.cs]: Can this be made into a const or something?

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ByteViewer.cs [hunk 13]
```diff
@@ -0,0 +1,753 @@
...
+            g.DrawString(hexString, font, foreground, ADDRESS_START_X, LINE_START_Y + line * CELL_HEIGHT);
+        }
+
+        /// <summary>
+        ///     Draws the client background and frames
+        /// </summary>
+        /// <internalonly/>
+        private void DrawClient(Graphics g)
+        {
+            using (Brush brush = new SolidBrush(SystemColors.ControlLightLight))
```
Comment (zsd4yr) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ByteViewer.cs]: I think these usings can be simplified

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ByteViewer.cs [hunk 14]
```diff
@@ -0,0 +1,753 @@
...
+            for (int i = 0; i < lineBuffer.Length; i++)
+            {
+                c = Convert.ToChar(lineBuffer[i]);
+                if (CharIsPrintable(c))
+                {
+                    sb.Append(c);
+                }
+                else
+                {
+                    sb.Append('.');
```
Comment (zsd4yr) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ByteViewer.cs]: const for the dot please

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ByteViewer.cs [hunk 15]
```diff
@@ -0,0 +1,753 @@
...
+                if (i == _columnCount / 2 - 1)
+                {
+                    result.Append(" ");  //add one extra in the middle
+                }
+            }
+
+            using Brush foreground = new SolidBrush(ForeColor);
+            g.DrawString(result.ToString(), font, foreground, HEX_START_X + BORDER_GAP, LINE_START_Y + line * CELL_HEIGHT);
+
+            /* ISSUE a-gregka: If perf problem, could be done this way to eliminate drawing twice on repaint
```
Comment (zsd4yr) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ByteViewer.cs]: Should we just remove this commented-out code?

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ByteViewer.cs [hunk 16]
```diff
@@ -0,0 +1,753 @@
...
+        /// </summary>
+        public virtual DisplayMode GetDisplayMode()
+        {
+            return _displayMode;
+        }
+
+        // Stole this code from  XmlSanner       
+        private static int GetEncodingIndex(int c1)
+        {
+            switch (c1)
```
Comment (zsd4yr) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ByteViewer.cs]: I would prefer this look like the new c# 8 switch expression

```c#
private static int GetEncodingIndex(int c1)
    => c1 switch 
    {
        0x0000 => 1,
        0xFEFF => 2,
        0xFFFE => 3,
        0x3C00 => 4,
        0x3c00 => 5,
        0x003c => 6,
        0x3f00 => 7,
        0x003f => 8,
        0x3c3f => 9,
        0x786d => 10,
        0x4c6f => 11,
        0xa794 => 12,
        _ => 0,
    };
```

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ByteViewer.cs [hunk 17]
```diff
@@ -0,0 +1,753 @@
...
+        /// <summary>
+        ///  Resize handler for the control.
+        /// </summary>
+        protected override void OnLayout(LayoutEventArgs e)
+        {
+            // Must call this first since we might return
+            base.OnLayout(e);
+
+            int rows = (ClientSize.Height - 2 * (BORDER_GAP + INSET_GAP)) / CELL_HEIGHT;
+            if (rows >= 0 && rows != _rowCount)
```
Comment (zsd4yr) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ByteViewer.cs]: Maybe invert?

```c#
if (rows < 0 || rows == _rowCount)
{
    return;
}

_rowCount = rows;
```

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ByteViewer.cs [hunk 18]
```diff
@@ -0,0 +1,753 @@
...
+        }
+
+        /// <summary>
+        ///  Sets the byte array to be displayed in the viewer.
+        /// </summary>
+        public virtual void SetBytes(byte[] bytes)
+        {
+            if (bytes == null)
+            {
+                throw new ArgumentNullException("bytes");
```
Comment (zsd4yr) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ByteViewer.cs]: nameof(bytes)

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ByteViewer.cs [hunk 19]
```diff
@@ -0,0 +1,753 @@
...
+                    break;
+            }
+        }
+
+        /// <summary>
+        ///  Sets the file to be displayed in the viewer.
+        /// </summary>
+        public virtual void SetFile(string path)
+        {
+            FileStream currentFile = new FileStream(path, FileMode.Open, FileAccess.Read, FileShare.None);
```
Comment (zsd4yr) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ByteViewer.cs]: would a using for the FileStream be more appropraite and avoid the catch and close?

**Comment by zsd4yr** [comment]
Lots of cleanup stuff. The ones I would _really_ like to see are:
- private and const variable names conform to convention
- separate nested class to its own file

**Comment by RussKie** [comment]
I can do that too
________________________________

**Comment by RussKie** [comment]
I fixed a few things with SCA. All other clean up is welcome out of bound.
